clients/drcachesim/tests/allasm-repstr-basic-counts.templatex

```diff
@@ -0,0 +1,37 @@
+Adios world!
+---- <application exited with code 0> ----
+Basic counts tool results:
+Total counts:
+          14 total \(fetched\) instructions
+          14 total unique \(fetched\) instructions
+/* One for each movs after the first one. */
```

Comment:
Unfortunately, in the repstr case, we cannot avoid the pc entry added by each iteration of the stringop (hence the 4 non-fetched instrs here). All instrs in the expanded repstr seq are in a loop, so the pc entry will be output each time.

This is unlike scatter/gather expansion where there's no loop, and we could choose to not instrument mov.

Reply:
?  These statistics are on the *final* trace which must have an entry for each iteration of the loop: we can't skip them in the final trace.  Whether the raw trace has all of them or not is a different matter.  Xref #4914.

Reply:
> These statistics are on the final trace which must have an entry for each iteration of the loop:

Note that the `Total non-fetched instructions` stat is for instrs, not memrefs.

My understanding is that the final trace would contain only a single `rep movs` instr entry. The raw trace would contain multiple pc entries all pointing to the `rep movs` pc; all of these entries are initially marked as `TRACE_TYPE_INSTR_MAYBE_FETCH`). Later only one of these will be fetched while constructing the final trace [here](https://github.com/DynamoRIO/dynamorio/blob/dcbaca81b9816a40fddc2b0211475a7c5a9220a3/clients/drcachesim/tracer/raw2trace.h#L1035), and others will be marked as [`TRACE_TYPE_INSTR_NO_FETCH`](https://github.com/DynamoRIO/dynamorio/blob/c530f52e3797ece0ac023876b65e42a66657a8ad/clients/drcachesim/common/trace_entry.h#L164).

Reply:
The final trace will have a separate instruction entry for each iteration of the rep string loop.  This is the whole point for the existence of "non-fetched" instructions.

Reply:
The point it, it's a conscious decision to have a separate instruction entry for each iteration of a rep string loop.  It is unrelated to how the raw trace is recorded or whether the instrumentation can easily avoid an instruction entry: post-processing could strip them all out if we wanted.  The final trace can look how we want it to look.

It also seems like a scatter/gather is basically a loop: it is not an atomic instruction.  It has an iteration over the mask bits.  If interrupted in the middle, it does not go back and re-execute the original instruction: it resumes in the middle with whatever last mask bit it finished with.

Reply:
Note that we ended up with mixed feelings about the non-fetched instructions and whether they should be there or not.

For scatter-gather: is the expansion uniform?  Could it be re-written as a loop?  Would that shrink it significantly?

Reply:
Thanks for clarifying. I think I misunderstood what "non fetched instrs" meant (I see now that they're NOT "instr entries that are ignored while making the final trace"). I verified using `opcode_mix` and `view` that there are indeed 5 instances of `rep movs` in the final trace.

> It also seems like a scatter/gather is basically a loop: it is not an atomic instruction. It has an iteration over the mask bits. If interrupted in the middle, it does not go back and re-execute the original instruction: it resumes in the middle with whatever last mask bit it finished with.

So I assume we want the same behaviour for scatter/gather as what we have for rep stringop? -- as many scatter/gather instrs in the final trace as the number of set bits.

Reply:
> So I assume we want the same behaviour for scatter/gather as what we have for rep stringop? -- as many scatter/gather instrs in the final trace as the number of set bits.

I'm not sure: even for rep string, for many cases we don't want the extra instructions.  E.g., perfctrs only count it as 1 instruction, and cache simulators should probably count it as 1 fetch.  Xref https://github.com/DynamoRIO/dynamorio/issues/2051#issuecomment-348601827.  Looking back, I kind of wish we had never made the "non fetched instrs" and had just one instr entry followed by just memrefs.

Reply:
Having a single instr and multiple memrefs makes sense to me -- that's the behaviour I assumed initially too. What was the rationale behind having multiple instances of the rep stringop in the instr stream?

Reply:
I think originally it was easiest to implement with an instr per iter, and that matched other cache simulators like Cachegrind and Pin's icache tool -- but I think both of those do similar rep string rewrites into explicit loops and maybe don't have good reasons to have a fetch per iter.  Only later did we look at it more closely with core simulations and comparing dr$sim to a wider range of simulators.  We seem to have assumed that a core simulator would want the extra instruction entries: https://github.com/DynamoRIO/dynamorio/issues/2051#issuecomment-348713878.  But, I'm not sure that's really true.  I suspect we could remove them and core simulators would be fine.  The issue now would be backward compatiblity: we'd have to label it a breakage.

Reply:
I see. So this PR can do the right thing and let there be one instr entry + following memref(s).

The rep stringop case will need to be fixed. I'll probably create a new issue, or add it to the i#2985 task list. As it's a breaking change, I'd prefer to do it in new PR atleast.

Reply:
Seems reasonable to file a new issue on re-considering "non-fetched instrs".  We would want to ask all concerned parties, including third parties running traces through simulators, for input to make sure it's an unused feature before removing it.

Reply:
> For scatter-gather: is the expansion uniform? 

Not sure what you mean by uniform?

> Could it be re-written as a loop? Would that shrink it significantly?

Initially I had thoughts of re-writing it as a loop, which would shrink the expansion size by a factor of number of store/load ops in the scatter/gather. 

But that may be independent of this discussion. To avoid multiple instr entries, we can find one instruction that's before the loop (in case of rep stringop, maybe we can add a dummy nop), and instrument that to output the pc entry.

Reply:
Though we may run into https://github.com/DynamoRIO/dynamorio/pull/4912#discussion_r633798543 if we instrument any non-app instr (like the dummy nop I suggested).

Reply:
> Not sure what you mean by uniform?

Oh you mean whether the instrs in each iteration are same so that it can be made into a loop. I haven't really done it yet, so not sure if there are any unknown barriers to doing that. I think we can do it independently of this PR. I'll add it to the i#2985 task list.

Reply:
Right.  A complication is that the looping backward branch must be an app instr for DR to make it into a loop.  We could augment drmgr_fix_app_ctis() or sthg.

Reply:
https://github.com/DynamoRIO/dynamorio/pull/4912#discussion_r634043217: reverted extra bools and switched to old logic for repstr detection and mitigation.

#4915 for changing behaviour for repstr to single instr entry.